src/time_tagger/connection/service.py

```diff
@@ -0,0 +1,110 @@
+from dataclasses import dataclass
+from shared.messages.ErrorMessageEnum import ErrorMessageEnum
+from time_tagger.connection.dao import ConnectionDao
+
+from time_tagger.connection.repository import (
+    ConnectionStatusEnum,
+    ConnectionRepository,
+    TimeTaggerConnectionInfoDto,
+)
+from time_tagger.hardware_properties.service import TimeTaggerHardwarePropertiesService
+
+
+@dataclass
+class TimeTaggerAddressInfoDto:
+    host_address: str
+    port: str
+    time_tagger_name: str
+
+
+@dataclass
+class TimeTaggerConnectionRes:
+    connected_devices: list[TimeTaggerConnectionInfoDto]
+    connection_failed_devices: list[TimeTaggerConnectionInfoDto]
+
+
+class ConnectionService:
+    def __init__(
+        self,
+        time_tagger_connection_data: ConnectionRepository,
+        time_tagger_hardware_service: TimeTaggerHardwarePropertiesService,
+        connection_dao: ConnectionDao,
+    ):
+        self.connection_data = time_tagger_connection_data
+        self.time_tagger_hardware_service = time_tagger_hardware_service
+        self.connection_dao = connection_dao
+
+    def connect_to_time_tagger_server(
+        self,
+        time_taggers_info: list[TimeTaggerAddressInfoDto],
+    ) -> TimeTaggerConnectionRes:
+
+        connected_devices = []
+        connection_failed_devices = []
+
+        for time_tagger in time_taggers_info:
+            try:
+                self._validate_time_tagger_name(time_tagger.time_tagger_name)
+
+                tagger = self.connection_dao.create_time_tagger_network_proxy(
+                    host_address=time_tagger.host_address, port=time_tagger.port
+                )
+
+                serial_number = self.time_tagger_hardware_service.get_time_tagger_serial_number(
+                    tagger
+                )
+
+                connection_info_res = self.connection_data.upsert_time_tagger_connection_info(
+                    TimeTaggerConnectionInfoDto(
+                        time_tagger_name=time_tagger.time_tagger_name,
+                        tagger_proxy=tagger,
+                        connection_status=ConnectionStatusEnum.SUCCESS.value,
+                        serial_number=serial_number,
+                    ),
+                )
```

Comment:
This is an interesting thing:
1. `connection_dao.create_time_tagger_network_proxy` is leaking an internal value (`tagger`)
2. we use the `tagger` as input to our `self.time_tagger_hardware_service.get_time_tagger_serial_number` so we get the `serial_number`
3. we add this `serial_number` value to `self.connection_data`, which also looks like a `dao` (maybe `connection_dao`?)
4. we store the connection info inside this service

For me this is doing a lot, which could be hidden inside the `connection_dao` and only return a `connection_info_res` at the end.

Also, we should consider moving this `time_tagger_hardware_service.get_time_tagger_serial_number` to the same `dao`, so we can reduce one level of dependency.

I'm not sure if we can do all of this. If so, let's consider having a follow-up PR to focus on changing this implementation.

Reply:
Good! I will create a task here to focus on this implementation later. I guess worth having a moment focused only on this